includes/REST_API/URL_Controller.php

```diff
@@ -0,0 +1,245 @@
+	}
+
+	/**
+	 * Registers routes for links.
+	 *
+	 * @since 1.11.0
+	 *
+	 * @see register_rest_route()
+	 *
+	 * @return void
+	 */
+	public function register_routes() {
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base,
+			[
+				[
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => [ $this, 'parse_url' ],
+					'permission_callback' => [ $this, 'parse_url_permissions_check' ],
+					'args'                => [
+						'url' => [
+							'description' => __( 'The URL to process.', 'web-stories' ),
+							'required'    => true,
+							'type'        => 'string',
+							'format'      => 'uri',
+						],
+					],
+				],
+			]
+		);
+	}
+
+	/**
+	 * Parses a URL to return some metadata for inserting links.
+	 *
+	 * @since 1.11.0
+	 *
+	 * @param WP_REST_Request $request Full data about the request.
+	 *
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function parse_url( $request ) {
+		$url = untrailingslashit( $request['url'] );
+
+		if ( empty( $url ) ) {
+			return new WP_Error( 'rest_invalid_url', __( 'Invalid URL', 'web-stories' ), [ 'status' => 404 ] );
+		}
+
+		/**
+		 * Filters the link data TTL value.
+		 *
+		 * @since 1.11.0
+		 *
+		 * @param int $time Time to live (in seconds). Default is 1 day.
+		 * @param string $url The attempted URL.
+		 */
+		$cache_ttl = apply_filters( 'web_stories_url_data_cache_ttl', DAY_IN_SECONDS, $url );
+		$cache_key = 'web_stories_url_data_' . md5( $url );
+
+		$data = get_transient( $cache_key );
+		if ( ! empty( $data ) ) {
+			$response = $this->prepare_item_for_response( json_decode( $data, true ), $request );
+
+			return rest_ensure_response( $response );
+		}
+
+		$response = wp_safe_remote_get( $url );
+
+		if ( is_wp_error( $response ) && 'http_request_failed' === $response->get_error_code() ) {
+			return new WP_Error( 'rest_invalid_url', __( 'Invalid URL', 'web-stories' ), [ 'status' => 404 ] );
+		}
+
+		$code = wp_remote_retrieve_response_code( $response );
+		if ( ! in_array( $code, [ WP_Http::OK, WP_Http::MOVED_PERMANENTLY, WP_Http::FOUND ], true ) ) {
+			return new WP_Error( 'rest_invalid_url', __( 'Invalid URL', 'web-stories' ), [ 'status' => 404 ] );
+		}
+
+		$headers = wp_remote_retrieve_headers( $response );
+
+		$mime_type          = $headers['content-type'];
+		$allowed_mime_types = $this->get_allowed_mime_types();
+		$allowed_mime_types = array_merge( ...array_values( $allowed_mime_types ) );
+
+		if ( ! in_array( $mime_type, $allowed_mime_types, true ) ) {
+			return new WP_Error( 'rest_invalid_file_type', __( 'File type', 'web-stories' ), [ 'status' => 404 ] );
+		}
+
+		$path = wp_parse_url( $url, PHP_URL_PATH );
+		$exts = $this->get_file_type_exts( [ $mime_type ] );
+		$ext  = end( $exts );
+
+		$file_name = basename( $path );
+		$data      = [
+			'ext'       => $ext,
+			'file_name' => $file_name,
+			'mime_type' => $mime_type,
+		];
```

Comment:
Let's also add something like `'size' => $headers['content-length'],` in case we might need this in the future